src/fsharp/service/IncrementalBuild.fs

```diff
@@ -1287,22 +1280,38 @@ type IncrementalBuilder(
         | Some slot -> slot
         | None -> failwith (sprintf "The file '%s' was not part of the project. Did you call InvalidateConfiguration when the list of files in the project changed?" filename)
 
-    member _.GetSlotsCount () = fileNames.Length
+    member _.GetSlotsCount () = mainState.sourceFiles.Length
 
     member this.ContainsFile(filename: string) =
         (this.TryGetSlotOfFileName filename).IsSome
 
     member builder.GetParseResultsForFile (filename) =
+        let state = currentState
         let slotOfFile = builder.GetSlotOfFileName filename
-        let fileInfo = fileNames.[slotOfFile]
+        let fileInfo = state.documents.[slotOfFile]
         // re-parse on demand instead of retaining
-        let syntaxTree = GetSyntaxTree fileInfo
+        let syntaxTree = GetSyntaxTree mainState fileInfo
         syntaxTree.Parse None
 
-    member _.SourceFiles  = sourceFiles  |> List.map (fun (_, f, _) -> f)
+    member _.SourceFiles  = mainState.sourceFiles |> Seq.map (fun (_, f, _) -> f) |> List.ofSeq
+
+    member this.UpdateDocuments(docs: FSharpDocument seq) =
```

Comment:
If two calls are made concurrently, one setting docA, another docB, then the update to one will be lost